### ISSUE TITLE
redirect URI

### ISSUE THREAD

**Issue by makushima** [Azure-Samples/ms-identity-java-desktop#21]
how to check redirect URI in Username-Password flow?

**Comment by Avery-Dunn** [comment]
Hello @makushima : This sample is meant to demonstrate the calling pattern for the [username/password flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-authentication-flows#usernamepassword-ropc), and doesn't need a redirect URI to retrieve tokens.

However, while username/password flow is supported it isn't recommended for applications accessed by normal users (it's more meant for simple back-end/devops scenarios). Depending on your scenario, I'd recommend checking out some other samples:
- If your users are [federated](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/whatis-fed), you can use the [integrated windows auth flow](https://github.com/Azure-Samples/ms-identity-msal-java-samples/tree/main/2.%20Client-Side%20Scenarios/Integrated-Windows-Auth-Flow) to sign in users and get tokens without them needing to provide their passwords
- If your users don't have access to a browser, [device code flow](https://github.com/Azure-Samples/ms-identity-msal-java-samples/tree/main/2.%20Client-Side%20Scenarios/Device-Code-Flow) will allow them to sign in on users using their phone or other device
- If you're developing a web application, we have tutorials showing how to sign in users and access Azure resources for both [Java servlet projects](https://github.com/Azure-Samples/ms-identity-msal-java-samples/tree/main/3.%20Java%20Servlet%20Web%20App%20Tutorial) and [projects using the Spring framework](https://github.com/Azure-Samples/ms-identity-msal-java-samples/tree/main/4.%20Spring%20Framework%20Web%20App%20Tutorial)

**Comment by ghost** [comment]
Hello, Avery-Dunn! We have third party Java app based on ADAL without source code. And as I understand it uses username password flow. But it also requires in app configuration Custom Redirect URI in format myapp://oauth. More details here https://www.oauth.com/oauth2-servers/redirect-uris/redirect-uris-native-apps/. I am not sure why they use this URI but it will be nice to have example/sample for backward compability

**Comment by Avery-Dunn** [comment]
@makushima Perhaps the redirect URI is being used as the authority in the client app? They are often the same or pretty similar, so the terminology is often used interchangeably.

In the sample, the section where the client app is created is where the authority/redirect URI is used:
```
PublicClientApplication pca = PublicClientApplication.builder(clientId)
                .authority(authority)  //<--- right here
                .build();
```

This value is configured in the `applications.properties` file with the `AUTHORITY` field. Normally username/password flow works just with the basic `https://login.microsoftonline.com/organizations/` we have in there already, but if your ADAL app was using a redirect URI in the username/password flow then that might be the equivalent in MSAL.

**Comment by ghost** [comment]
@Avery-Dunn No. I don't think that it is Authority. The authority is URL to Authentication Authority. And redirect URI is as the article says: 
_Native applications are clients installed on a device, such as a desktop application or native mobile application. There are a few things to keep in mind when supporting native apps related to security and user experience.

The authorization endpoint normally redirects the user back to the client’s registered redirect URL. Depending on the platform, native apps can either claim a URL pattern, or register a custom URL scheme that will launch the application. For example, an iOS application may register a custom protocol such as myapp:// and then use a redirect_uri of myapp://callback._

So old java ADAL app has redirect URI: vendorcorp://applicationname. If I change URI on Azure side, ADAL app gives error message and doesn't work. How can I check in sample this redirect URI?

**Comment by Avery-Dunn** [comment]
Since you mentioned that you had your app "without source code", are you sure that app was using the [username/password flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-authentication-flows#usernamepassword-ropc) ([ROPC](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth-ropc))?

I'm not as familiar with ADAL Java but just looking at where it [builds the acquire token call for the username/password flow](https://github.com/AzureAD/azure-activedirectory-library-for-java/blob/cf2fe74a35c0d7cc6cce24165bb9be47068b6bcb/src/main/java/com/microsoft/aad/adal4j/AuthenticationContext.java#L177) it doesn't seem like a redirect URI is used, nor can you send it [as a parameter](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/src/main/java/com/microsoft/aad/msal4j/UserNamePasswordParameters.java) in MSAL Java's username/password flow.

Maybe the old app was actually using the [authorization code flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-authentication-flows#authorization-code)? Unlike the username/password flow where user credentials are either hardcoded or retrieved from users through your own login page, the auth code flow is meant for web applications which redirects users to another place to enter their credentials (for example https://login.microsoftonline.com/).

Have a look at [our web app sample](https://github.com/Azure-Samples/ms-identity-java-webapp/tree/master/msal-java-webapp-sample) and see if it better matches what the app that used ADAL did. In that sample, [here](https://github.com/Azure-Samples/ms-identity-java-webapp/blob/b37376b1fc481704b9b96b77170f01392f44bcaa/msal-java-webapp-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthFilter.java#L40) is where it builds a redirect URI, and [here](https://github.com/Azure-Samples/ms-identity-java-webapp/blob/b37376b1fc481704b9b96b77170f01392f44bcaa/msal-java-webapp-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthHelper.java#L177) is where it uses that redirect URI to get an auth code (that is later used to get the access token).

The only other place I can think of where a redirect URI would be used is an [interactive request where MSAL opens the operating system's default browser](https://github.com/AzureAD/microsoft-authentication-library-for-java/wiki/Acquiring-Token-Interactively), however I don't think there was an ADAL equivalent for that auth flow and it's a pretty niche option so I'd be surprised if the old ADAL app was using it.

**Comment by ghost** [comment]
Yes, we know that vendor used Android SDK with ADAL ROPC. Also it is not webapp and redirect URL is not http://something. Redirect URL has structure microsoft://outlook format (see article for native app). The app authenticates non interactively, like service on Windows.

**Comment by Avery-Dunn** [comment]
Oh, if the app was on Android was it using [ADAL for Android](https://github.com/AzureAD/azure-activedirectory-library-for-android) rather than [ADAL for Java](https://github.com/AzureAD/azure-activedirectory-library-for-java)? If so and you're still making apps for Android then you'll probably want [MSAL Android](https://github.com/AzureAD/microsoft-authentication-library-for-android), but if it was using ADAL Android and you want to migrate to MSAL Java then there might not be direct translation for how redirect URIs were used.
